### PR TITLE
Update overview page

### PR DESCRIPTION
**Pull request by U8NWXD**

### CODE REVIEW

**Comment by U8NWXD** [comment]
@DubeySandeep this PR doesn't wrap lines. lmk me know if this causes any issues with reviewing

**Comment by DubeySandeep** [comment]
@vojtechjelinek can you take a pass on this PR? (Asking you as you have recently made changes to the codebase structure!)

File: Overview-of-the-Oppia-codebase.md [hunk 1]
```diff
@@ -1,57 +1,322 @@
...
+
+These classes don't just store information. They also provide `from_dict` and `to_dict` methods to convert between classes and dictionaries (which can be converted to JSON), and they can contain validation and converters to convert between different versions of the model.
+
+The domain models are defined in `*_domain.py` files under `core/domain/`.
+
+### Storage layer
+
+Finally, we have the storage layer, which defines the storage models.  These are also classes, but they define how data is stored in whatever system we are using to store data to the file system. In production, we use the [Google Cloud Datastore](https://cloud.google.com/datastore), and we interface with it from Python using [Cloud NDB](https://googleapis.dev/python/python-ndb/latest/index.html). (Cloud NDB is for Python 3. For Python 2, we use App Engine NDB.)
+
+Ideally we'd like all services to access the datastore through the domain layer, but some services access the storage layer directly. While you shouldn't write any new code like this, you might run into old code that works this way.
```
Comment (vojtechjelinek) [inline Overview-of-the-Oppia-codebase.md:137]: This is a bit confusing, it is completely normal to interact with the storage through services, we just want to generate the domain object from the datastore class as soon as possible.

File: Overview-of-the-Oppia-codebase.md [hunk 2]
```diff
@@ -1,57 +1,322 @@
...
+
+The storage models are defined in `core/storage`, while the code that handles interacting with the datastore is in `core/platform`.  `core/platform/models.py` provides an interface to these storage models and the underlying datastore that dynamically loads the correct classes to interface with whatever datastore is currently in use. All the code in the controller, service, and domain layers should remain platform-agnostic by relying on the storage layer.
+
+For example, we use different underlying data storage platforms in dev mode (when running locally on developer machines) versus in prod mode (when running in production). Switching between these platforms is handled by the code in `core/platform`, so the rest of the backend layers can ignore them.
```
Comment (vojtechjelinek) [inline Overview-of-the-Oppia-codebase.md:141]:
```suggestion
For example, we use different underlying file storage platforms in dev mode (when running locally on developer machines) versus in prod mode (when running in production). Switching between these platforms is handled by the code in `core/platform`, so the rest of the backend layers can ignore them.
```

**Review by vojtechjelinek — CHANGES_REQUESTED** [review]
Thanks! Few small notes.

File: Overview-of-the-Oppia-codebase.md (hunk 2 above)
Comment (U8NWXD) [reply Overview-of-the-Oppia-codebase.md]: done

File: Overview-of-the-Oppia-codebase.md (hunk 1 above)
Comment (U8NWXD) [reply Overview-of-the-Oppia-codebase.md]: Thanks for clarifying. I've deleted this

**Comment by U8NWXD** [comment]
Thanks for the feedback @vojtechjelinek! I've addressed your comments. PTAL

**Review by vojtechjelinek — APPROVED** [review]
LGTM!

**Comment by U8NWXD** [comment]
@DubeySandeep do you want to take a look, or should I merge it now that vojtechjelinek has approved?

**Comment by DubeySandeep** [comment]
@vojtechjelinek Thanks for the review! @U8NWXD Feel free to merge this PR and update the wiki page! :)

**Pull request merged**